router/pkg/rulerouter/rulerouter.go

```
@@ -249,4 +266,33 @@ func (r *RuleRouterImpl) Config() *config.Router {
 	return r.rcfg
 }
 
+func (r *RuleRouterImpl) AddClient(cl rclient.RouterClient) {
+	r.clmu.Lock()
+	defer r.clmu.Unlock()
+	r.clmp[cl.GetCancelPid()] = cl
+}
+
+func (r *RuleRouterImpl) ReleaseClient(cl rclient.RouterClient) {
+	r.clmu.Lock()
+	defer r.clmu.Unlock()
+	delete(r.clmp, cl.GetCancelPid())
+}
+
+func (r *RuleRouterImpl) CancelCleint(csm *pgproto3.CancelRequest) error {
+	r.clmu.Lock()
+	defer r.clmu.Unlock()
+
+	if cl, ok := r.clmp[csm.ProcessID]; ok {
+		if cl.GetCancelKey() != csm.SecretKey {
+			return fmt.Errorf("cancel sekret does not match")
+		}
+		if cl.Server() != nil {
+			spqrlog.Logger.Printf(spqrlog.DEBUG1, "cancelling client pid %d", csm.ProcessID)
+			return cl.Server().Cancel()
+		}
+		return nil
+	}
+	return fmt.Errorf("no client with pid %d", csm.ProcessID)
```

Im not sure if we should return an error here. Is this error will be passing to the client?
Seems like if there is no process, then client was already released?

---

this error will not be passed to the client